cdap-formats/src/main/java/co/cask/cdap/format/io/JsonStructuredRecordDatumReader.java

```diff
@@ -0,0 +1,161 @@
+      .put(Schema.Type.BYTES, JsonToken.BEGIN_ARRAY)
+      .put(Schema.Type.ARRAY, JsonToken.BEGIN_ARRAY)
+      .put(Schema.Type.MAP, JsonToken.BEGIN_OBJECT)
+      .put(Schema.Type.RECORD, JsonToken.BEGIN_OBJECT)
+      .build()
+  );
+
+  @Override
+  public StructuredRecord read(Decoder decoder, Schema sourceSchema) throws IOException {
+    if (!(decoder instanceof JsonDecoder)) {
+      throw new IOException("The JsonStructuredRecordDatumReader can only decode using a JsonDecoder");
+    }
+
+    return super.read(decoder, sourceSchema);
+  }
+
+  @Override
+  protected String decodeEnum(Decoder decoder, Schema schema) throws IOException {
+    return getJsonReader(decoder).nextString();
+  }
+
+  @Override
+  protected Collection<?> decodeArray(Decoder decoder, Schema elementSchema) throws IOException {
+    List<Object> array = new ArrayList<>();
+
+    JsonReader jsonReader = getJsonReader(decoder);
+    jsonReader.beginArray();
+    while (jsonReader.peek() != JsonToken.END_ARRAY) {
+      array.add(decode(decoder, elementSchema));
+    }
+    jsonReader.endArray();
+
+    return array;
+  }
+
+  @Override
+  protected Map<?, ?> decodeMap(Decoder decoder, Schema keySchema, Schema valueSchema) throws IOException {
+    if (!keySchema.isCompatible(Schema.of(Schema.Type.STRING))) {
+      throw new IOException("Complex key type not supported: " + keySchema);
+    }
+
+    Map<Object, Object> result = new HashMap<>();
+
+    JsonReader jsonReader = getJsonReader(decoder);
+
+    jsonReader.beginObject();
+    while (jsonReader.peek() != JsonToken.END_OBJECT) {
+      Object key = convertKey(jsonReader.nextName(), keySchema.getType());
+      result.put(key, decode(decoder, valueSchema));
+    }
+    jsonReader.endObject();
+
+    return result;
+  }
+
+  @Override
+  protected StructuredRecord decodeRecord(Decoder decoder, Schema schema) throws IOException {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+
+    JsonReader jsonReader = getJsonReader(decoder);
+
+    jsonReader.beginObject();
+    while (jsonReader.peek() != JsonToken.END_OBJECT) {
+      Schema.Field field = schema.getField(jsonReader.nextName());
+      if (field == null) {
+        // Ignore unrecognized fields
+        jsonReader.skipValue();
+        continue;
+      }
+
+      builder.set(field.getName(), decode(decoder, field.getSchema()));
+    }
+    jsonReader.endObject();
+
+    return builder.build();
+  }
+
+  @Override
+  protected Object decodeUnion(Decoder decoder, Schema unionSchema) throws IOException {
+    JsonReader jsonReader = getJsonReader(decoder);
+    JsonToken token = jsonReader.peek();
+    // Based on the token to guess the schema
+    for (Schema schema : unionSchema.getUnionSchemas()) {
+      if (SCHEMA_TO_JSON_TYPE.get(schema.getType()) == token) {
+        return decode(decoder, schema);
+      }
+    }
+
+    throw new IOException("No matching schema found for union type: " + unionSchema);
```

Comment:
better to include the token that we couldn't find the match for in the error message